libraries/browser-client/src/client/browser_client.ts

```diff
@@ -0,0 +1,278 @@
+      // if (this.isWebkit) {
+      //   // Multiply base buffer size of 4 kB by the resample ratio rounded up to the next power of 2.
+      //   // i.e. for 48 kHz to 16 kHz downsampling, this will be 4096 (base) * 4 = 16384.
+      //   const resampleRatio = this.audioContext.sampleRate / this.sampleRate
+      //   const bufSize = baseBufferSize * Math.pow(2, Math.ceil(Math.log(this.resampleRatio) / Math.log(2)))
+      //   this.audioProcessor = this.audioContext.createScriptProcessor(bufSize, 1, 1)
+      // } else {
+      //   this.audioProcessor = this.audioContext.createScriptProcessor(undefined, 1, 1)
+      // }
+      // this.audioContext.createMediaStreamSource(this.mediaStream).connect(this.audioProcessor)
+      // this.audioProcessor.connect(this.audioContext.destination)
+      // this.audioProcessor.addEventListener(audioProcessEvent, (event: AudioProcessingEvent) => {
+      //   this.handleAudio(event.inputBuffer.getChannelData(0))
+      // })
+    }
+    await this.decoder.setSampleRate(this.audioContext?.sampleRate)
+    this.initialized = true
+  }
+
+  async close(): Promise<void> {
+    await this.detach()
+    this.speechlyNode?.port.close()
+    this.speechlyNode?.disconnect()
+    await this.decoder.close()
+    // Disconnect and stop ScriptProcessorNode
+    // if (this.audioProcessor != null) {
+    //   const proc = this.audioProcessor
+    //   proc.disconnect()
+    // }
+  }
+
+  async attach(mediaStream: MediaStream): Promise<void> {
+    await this.initialize()
+    await this.audioContext?.resume()
+
+    if (!this.speechlyNode) {
+      throw Error('cannot attach to stream, no node')
+    }
+    this.stream = this.audioContext?.createMediaStreamSource(mediaStream)
+    this.stream?.connect(this.speechlyNode)
+  }
+
+  async detach(): Promise<void> {
+    if (this.active) {
+      await this.stop()
+    }
+    if (this.stream) {
+      this.stream.disconnect()
+      this.stream = undefined
+    }
+  }
+
+  async uploadAudioData(audioData: ArrayBuffer): Promise<void> {
+    await this.initialize()
+    const audioBuffer = await this.audioContext?.decodeAudioData(audioData)
+    if (audioBuffer === undefined) {
+      throw Error('Could not decode audioData')
+    }
+    const samples = audioBuffer.getChannelData(0)
+
+    // convert 2-channel audio to 1-channel if need be
+    if (audioBuffer.numberOfChannels > 1) {
+      const chan1samples = audioBuffer.getChannelData(1)
+      for (let i = 0; i < samples.length; i++) {
+        samples[i] = (samples[i] + chan1samples[i]) / 2.0
+      }
+    }
+
+    await this.start()
+
+    let sendBuffer: Float32Array
+    for (let b = 0; b < samples.length; b += 16000) {
+      const e = b + 16000
+      if (e > samples.length) {
+        sendBuffer = samples.slice(b)
+      } else {
+        sendBuffer = samples.slice(b, e)
+      }
+      this.handleAudio(sendBuffer)
+    }
+
+    await this.stop()
+  }
+
+  async start(): Promise<string> {
```

Comment:
What if I call this without attaching a MediaStream?

Reply:
Also, I believe we should be able to pass an `appId` (project based login) as well as other parameters here... Now I'm mostly thinking of the shallow fusion vocabulary.

Reply:
If start is called without attach, then no data would be sent. Maybe the uploader could be a separate class, then the MediaStream support could demand an attached MediaStream when calling start.